drivers/serial/uart_sifive.c

```diff
@@ -8,6 +8,8 @@
  * @brief UART driver for the SiFive Freedom Processor
  */
 
+#define DT_DRV_COMPAT sifive_uart0
```

Comment:
Any way we can we use the real compatible string ("sifive,uart0") here ?
If not, I'd actually prefer the caps version. Adding a new, intermediate (with "_" but lower case) modified string is confusing IMHO.

Reply:
I believe this must be lexically a C identifier fragment so it can be part of a macro identifier, so using a comma is out.

AFAICT the only SiFive-related compatible that is not consistently lower-case is the SOC `SiFive,FE310G-0002-Z0-soc`. which IMO is incorrectly named not least because it's not the same as the other SiFive compatibles.

I'd support preserving case in the compatible string as long as the devicetree standard and/or reference implementation dtc considers case significant when checking compatibility.  If non-Zephyr devicetree considers case insignificant then I'd support canonicalizing to the lower-case.

Reply:
As long as we don't use the exact string, I don't see the point to keep the exact same case.
So, from the moment we have to convert the string, I'd say `SIFIVE_UART0` would be easier to grep (and potentially find related generated macros)

Reply:
> Any way we can we use the real compatible string ("sifive,uart0") here ?

I definitely don't have the preprocessor skills to make that happen. Not sure it's even possible, really.

The generated header looks like this:

```
#define DT_N_INST_0_sifive_uart0 DT_N_S_soc_S_serial_10013000

/* Generic property macros: */
#define DT_N_S_soc_S_serial_10013000_P_reg {268513280 /* 0x10013000 */, 4096 /* 0x1000 */}
#define DT_N_S_soc_S_serial_10013000_P_reg_EXISTS 1
#define DT_N_S_soc_S_serial_10013000_P_interrupts {3 /* 0x3 */}
#define DT_N_S_soc_S_serial_10013000_P_interrupts_EXISTS 1
#define DT_N_S_soc_S_serial_10013000_P_clock_frequency 16000000
#define DT_N_S_soc_S_serial_10013000_P_clock_frequency_EXISTS 1
... etc
```

So the macro expansion is roughly `DT_DRV_INST(0)` -> `DT_INST(0, DT_DRV_COMPAT)` -> `DT_INST(0, sifive_uart0)` -> `DT_N_INST_0_sifive_uart0` -> `DT_N_S_soc_S_serial_10013000`.

The final token `DT_N_S_soc_S_serial_10013000` can get pasted onto `_P_clock_frequency` or something like that to get property values, etc.

So we're relying on being able to paste DT_DRV_COMPAT in the preprocessor, like `DT_N_INST_##inst##_##DT_DRV_COMPAT`, for this all to work. That means it has to be a C token.

So using "sifive,uart0" would require a solution where we could convert that string into that token at preprocessor time. I don't know how to do that.... any ideas?

> If not, I'd actually prefer the caps version. Adding a new, intermediate (with "_" but lower case) modified string is confusing IMHO.

The underlying macros require that names from devicetree are lowercased in order to avoid possible namespace clashes. Reversing that so they are uppercase is possible, but would be a lot of work at this point. Is this a strong preference?

Reply:
This is the same requirement introduced for labels, properties, etc, btw -- special characters are consistently converted to underscores everywhere.

Reply:
> Reversing that so they are uppercase is possible, but would be a lot of work at this point. Is this a strong preference?

To clarify, we would do this everywhere: properties would also need to be uppercased, etc. And the generated macro names would have to use lowercase, so it would look like `DT_n_s_SOC_s_SERIAL_10013000_p_CLOCK_FREQUENCY` in the generated file.

IMO this is a worse option because devicetree identifiers tend to be lowercase most of the time, so people just need to know to convert special characters to underscores in the C API.

We do need a consistent case in this naming scheme to avoid collisions. For example, uppercase _P_ is used to separate the node identifier from the property name. Since the identifier and property names are lowercase, there is no ambiguity. If we don't do a case transformation, we can't be sure.

Please see macros.bnf for the details.

Reply:
@mbolivar-nordic, thanks for the detailed explanation. Ack to say it's minor a drawback compared to what the global solution conveys

Reply:
@mbolivar-nordic, thanks for the detailed explanation. Ack to say it's minor a drawback compared to what the global solution conveys